Disable Crash Reporting Draft (#2769)
MicrosoftEdge/WebView2Feedback: specs/DisableCrashReporting.md

Process Info
===

# Background
Currently, when a WebView2 is created a WebView2 crashpad process will be created in addition to starting a WebView2 browser process. If any WebView2 process crashes, the crashpad process will create crash dumps and send them to Microsoft for diagnosis. This document covers new APIs to allow the end developer to customize crash reporting to help when running diagnostics and doing analysis. They can set the `CoreWebView2EnvironmentOptions.IsCustomCrashReportingEnabled` property to not have crash dumps sent to Microsoft and use the `CoreWebView2Environment.CrashDumpFolderPath` property to locate crash dumps and do customization with them instead.

# Examples
## WinRT and .NET   
```c#

/// Create WebView Environment with option
void CreateEnvrionmentWithOption()
{
    CoreWebView2EnvironmentOptions options = new CoreWebView2EnvironmentOptions();
    options.IsCustomCrashReportingEnabled = true;
    CoreWebView2Environment environment = await CoreWebView2Environment.CreateAsync(BrowserExecutableFolder, UserDataFolder, options);
}

string _crashDumpFolder;

void GetDumpFolder()
{

>     _crashDumpFolder = webView.CoreWebView2.Environment.CrashDumpFolderPath;

Reviewer: We probably don't need to store this in a member variable but just get it when we need it like we do in the C++ sample?

}

void WebView_CoreWebView2InitializationCompleted(object sender, CoreWebView2InitializationCompletedEventArgs e)
{
    webView.CoreWebView2.ProcessFailed += WebView_ProcessFailed;
}

void WebView_ProcessFailed(object sender, CoreWebView2ProcessFailedEventArgs e)
{
    // When process failed, do custom parsing with dumps
    ProcessNewCrashDumps(_crashDumpFolder);
}

```
## Win32 C++
```cpp
void AppWindow::InitializeWebView()
{
    auto options = Microsoft::WRL::Make<CoreWebView2EnvironmentOptions>();
    CHECK_FAILURE(options->put_IsCustomCrashReportingEnabled(TRUE));
    // ... other option properties

    // ... CreateCoreWebView2EnvironmentWithOptions
}

void AppWindow::ProcessFailed()
{
    // Get dump folder path
    wil::com_ptr<ICoreWebView2Environment11> environment;
    wil::unique_cotaskmem_string crashDumpFolder;
    CHECK_FAILURE(environment->get_CrashDumpFolderPath(&crashDumpFolder));
	
	// Register a handler for the ProcessFailed event.
    CHECK_FAILURE(m_webView->add_ProcessFailed(
        Callback<ICoreWebView2ProcessFailedEventHandler>(

>             [this](ICoreWebView2* sender, ICoreWebView2ProcessFailedEventArgs* argsRaw)

Reviewer: you aren't capturing crashDumpFolder from above for this lambda. Does this work?

Author: Forgot to capture. Updated

>             [this, &crashDumpFolder](ICoreWebView2* sender, ICoreWebView2ProcessFailedEventArgs* argsRaw)

Reviewer: Does capturing with '&' work? Won't that get a reference to the local variable object which will go out of scope and then we'll crash in the event handler? I think you want just crashDumpFolder without the '&'?

Process Info
===

# Background

> Currently, the crash process is created asynchronously with the WebView2 process and listening for any crashes. Once a crash happens, the crash pad process will create a crash dump and send that to Microsoft services for diagnosis. Provide end developer a new API to toggle dump upload consent, once uploads are disabled, dump will still be generated but just not uploaded to Microsoft endpoints. Also, provide an API to get the crash dump folder.

Reviewer: Throughout the doc please make sure line lengths don't go beyond 80 or 100 columns. Its difficult for reviewers to read in github UI

> Currently, the crash process is created asynchronously with the WebView2 process and listening for any crashes. Once a crash happens, the crash pad process will create a crash dump and send that to Microsoft services for diagnosis. Provide end developer a new API to toggle dump upload consent, once uploads are disabled, dump will still be generated but just not uploaded to Microsoft endpoints. Also, provide an API to get the crash dump folder.

Reviewer: ```suggestion
Currently, when a WebView2 is created a WebView2 crashpad process will be created in addition to starting a WebView2 browser process. If any WebView2 process crashes, the crashpad process will create crash dumps and send them to Microsoft for diagnosis. This document covers new APIs to allow the end developer to customize crash reporting. They can set the `CoreWebView2EnvironmentOptions.IsCustomCrashReportingEnabled` property to not have crash dumps sent to Microsoft and use the `CoreWebView2Environment.CrashDumpFolderPath` property to locate crash dumps and do whatever they like with them instead.
```

# Examples
## WinRT and .NET   
```c#

> /// Create WebView Environment with option

Reviewer: This sample is close to a real end to end scenario demonstration. If an end dev wants to do custom crash reporting, they should disable our crash reporting and then when there's a crash, get the files out of the crash dump folder and do something with them. We don't have to show doing something with the crash dumps (we can call a function like `ProcessNewCrashDumps(crashDumpFolder);` that we don't define) but we should show how they should know when to find new crash dumps. Can they use the ProcessFailed event - do we guarantee that the crash dumps will be totally written to disk by the time the ProcessFailed event is raised? Or do they have to do polling or some other way of watching for changes to the dump folder?

Author: The most reliable way for end developer is to monitor crash dump folder's file changes. (which is what DDG do)

Author: I verified using debugger. Dumps will get write to disk when ProcessFailed are raised. I updated the sample code section

auto options = new CoreWebView2EnvironmentOptions();
options.DisableCrashReporting = true;
auto environment = await CoreWebView2Environment.CreateAsync(BrowserExecutableFolder, UserDataFolder, options);

/// Webview creation and setup of completion handler are out of scope of sample
void WebView_CoreWebView2InitializationCompleted(object sender, CoreWebView2InitializationCompletedEventArgs e)
{
    if (e.IsSuccess)
    {
        // Get the crash dump folder

>         String crashDumpFolder = webView.CoreWebView2.Environment.CrashDumpFolder();

Reviewer: C# properties are just '.CrashDumpFolder' no '()' at the end.

    }
}

```
## Win32 C++
```cpp

> void AppWindow::ToggleCrashReportingUploadConsent()

Reviewer: The C++ & C# samples should match.

{
    // Disable crash report
    auto options = Microsoft::WRL::Make<CoreWebView2EnvironmentOptions>();
    CHECK_FAILURE(options->put_DisableCrashReporting(TRUE));
}

void AppWindow::GetCrashDumpFolder()
{
    // Get crashpad folder path  
    wil::com_ptr<ICoreWebView2Environment11> environment;
    wil::unique_cotaskmem_string crashPadPath;
    CHECK_FAILURE(environment->get_CrashDumpFolder(&crashPadPath));
}
```

# API Details    
```
interface ICoreWebView2Environment11;
interface ICoreWebView2EnvironmentOptions3;

/// A continuation of the ICoreWebView2Environment interface for
/// getting the crash dump folder path
[uuid(F619312E-0399-4520-B700-30818441785A), object, pointer_default(unique)]
interface ICoreWebView2Environment11 : ICoreWebView2Environment10 {
  /// Get the crash dump folder that all CoreWebView2's from this

>   /// environment are using.

Reviewer: We need more information on what a crash dump folder is. Maybe something like

```suggestion
  /// environment are using. Whenever a WebView2 process crashes, a crash dump file will be created in the crash dump folder.
```

>   /// environment are using.

Reviewer: Similar to example code, we should document the recommendation of when to check the folder for new dumps. Can they use ProcessFailed?

Author: Updated the comment section

>   // MSOWNERS: xiaqu@microsoft.com

Reviewer: Somewhere we need documentation about how long crash dumps are kept, how many are kept. Does that documentation already exist and you can link to it from here?

Author: Let me find out if there are public documentations on this

Reviewer 2: We don't have anything today, but we should. I'll follow-up to get Diana the right contacts.

>   // MSOWNERS: xiaqu@microsoft.com

Reviewer: It would also be good to say what the crash dump format is. Maybe MSDN already has documentation on this (is it a minidump file? if so https://docs.microsoft.com/en-us/windows/win32/debug/minidump-files)

Author: Verified with reliability folks, minidump is the format. Should I document both in `CrashDumpFolderPath` API and `IsCustomCrashReportingEnabled` API?

Reviewer: Just the CrashDumpFolderPath where they need to know what is in that folder.

  [propget] HRESULT CrashDumpFolder([out, retval] LPWSTR* value);
}

/// Additional options used to create WebView2 Environment.
[uuid(3FB94506-58AB-4171-9082-E7D683471A48), object, pointer_default(unique)]
interface ICoreWebView2EnvironmentOptions3 : ICoreWebView2EnvironmentOptions2 {

  /// When `DisableCrashReporting` is set to `TRUE`, Windows won't send crash data to Microsoft endpoint.
  /// `DisableCrashReporting` is default to be `FALSE`, in this case, WebView respect OS consent. 
  // MSOWNERS: xiaqu@microsoft.com
  [propget] HRESULT DisableCrashReporting([out, retval] BOOL* value);

  /// Sets the `DisableCrashReporting` property.
  // MSOWNERS: xiaqu@microsoft.com
  [propput] HRESULT DisableCrashReporting([in] BOOL value);
}
```

```c# (but really MIDL3)
namespace Microsoft.Web.WebView2.Core
{
    
    // ...
    runtimeclass CoreWebView2EnvironmentOptions
    {
        // ...

>         Boolean DisableCrashReporting { get; set; };

Reviewer: How about `IsCustomCrashReportingEnabled`?

    }

    runtimeclass CoreWebView2Environment
    {

>         String CrashDumpFolder { get; };

Reviewer: We try to use '...Path' suffix for property names that return strings that are Windows file paths. 
```suggestion
        String CrashDumpFolderPath { get; };
```

Author: Is this c# specific? Or same change for C++ code

Reviewer: Same for COM & .NET

Process Info
===

# Background

> Currently, when a WebView2 is created a WebView2 crashpad process will be created in addition to starting a WebView2 browser process. If any WebView2 process crashes, the crashpad process will create crash dumps and send them to Microsoft for diagnosis. This document covers new APIs to allow the end developer to customize crash reporting. They can set the `CoreWebView2EnvironmentOptions.IsCustomCrashReportingEnabled` property to not have crash dumps sent to Microsoft and use the `CoreWebView2Environment.CrashDumpFolderPath` property to locate crash dumps and do customization with them instead.

Reviewer 3: Suggestion for adding to the third sentence: "...to customize crash reporting to help when running diagnostics and doing analysis."

# Examples
## WinRT and .NET   
```c#

/// Create WebView Environment with option
void CreateEnvrionmentWithOption()
{
    CoreWebView2EnvironmentOptions options = new CoreWebView2EnvironmentOptions();
    options.IsCustomCrashReportingEnabled = true;
    CoreWebView2Environment environment = await CoreWebView2Environment.CreateAsync(BrowserExecutableFolder, UserDataFolder, options);
}

void DirFileChanged()
{
    String crashDumpFolder = webView.CoreWebView2.Environment.CrashDumpFolderPath;
    ProcessNewCrashDumps(crashDumpFolder);
}

```
## Win32 C++
```cpp
void AppWindow::InitializeWebView()
{
    auto options = Microsoft::WRL::Make<CoreWebView2EnvironmentOptions>();
    CHECK_FAILURE(options->put_IsCustomCrashReportingEnabled(TRUE));
    // ... other option properties

    // ... CreateCoreWebView2EnvironmentWithOptions
}

> void AppWindow::DirFileChanged()

Reviewer: In both of these samples can we show how to hook up DirFileChanged? How does that method get called?

Author: Here is some samples https://docs.microsoft.com/en-us/windows/win32/fileio/obtaining-directory-change-notifications. But do we want to write the whole thing in our sample code and sample app?

Reviewer 3: Agreed here, we should show an example of how to use the DirFileChanged API. Can we do what Dave recommended in the comment above? https://github.com/MicrosoftEdge/WebView2Feedback/pull/2769/files#r965318889

Author: DirFileChanged is not an WebView API. I am still verifying on when dump file are written to disk

Reviewer 3: Sorry, I meant using the API to get the crash dump folder.
